project/SUNGOD3/README.rst

```diff
@@ -0,0 +1,154 @@
+   specific RAG applications.
+3. Creating a flexible, extensible framework that allows for easy 
+   experimentation with different search algorithms and metrics.
+4. Balancing speed, accuracy, and memory usage in vector search operations.
+
+Prospective Users
+-----------------
+This project will benefit:
+
+1. Researchers and developers working on RAG systems.
+2. Data scientists and machine learning engineers dealing with large-scale 
+   similarity search problems.
+3. Information retrieval specialists looking for customizable vector 
+   search solutions.
+4. Anyone working with high-dimensional data requiring efficient similarity 
+   search capabilities.
+5. Graduate students who lack a graduation thesis.
+
+System Architecture
+-------------------
+The AdvancedVectorSearch system will consist of the following components:
+
+1. Dataset Module: For generating and managing large-scale, high-dimensional 
+   vector datasets for testing and benchmarking.
+2. Index Module: Implementing various similarity comparison functions and 
+   indexing structures.
+3. Search Module: Providing add and search functionalities with support for 
+   different search algorithms.
+4. Benchmark Module: For comparing the implemented methods against linear 
+   search and potentially FAISS in terms of speed, accuracy, and memory usage.
+
+Index Module
+------------
+
+The Index Module is responsible for implementing various similarity comparison 
+functions and indexing structures. It supports basic operations for adding, 
+searching, and removing vectors from the index. The basic index structure 
+includes the following:
+
+- **vector<float>** : Stores the vector data.
+- **int** : Stores the vector dimension.
+- **int** : Stores the number of vectors in the index.
+- **string** : Unique identifier for each vector.
+
+The module provides the following methods:
+
+- **add(vector)** : Adds vectors to the index.
+- **search(query_vector)** : Searches the index given a query vector.
+- **remove(vector)** : Removes vectors from the index.
+
+Supported Index Types
+---------------------
+
+The Index Module supports the following types of indices:
+
+1. **Flat Index**: 
+   A simple index that stores vectors in memory and performs an 
+   exhaustive search.
+
+2. **IVF (Inverted File) Index**: 
+   An index that partitions the vector space into clusters for 
+   faster search.
+
+   This method provides a more efficient search process compared 
+   to flat indices, especially for large-scale datasets.
+
+The system will be implemented in C++ for core functionalities, with Python 
+bindings for ease of use.
+
+API Description
+---------------
+
+The AdvancedVectorSearch API will include:
+
+1. Dataset Management
+2. Index Operations
+3. Search Operations
+4. Benchmark
+
+Engineering Infrastructure
+--------------------------
+
+- Version Control: Git, hosted on GitHub
+- CI: GitHub Actions for automated testing and benchmarking
```

Comment:
[Question]
How do you benchmarking your project in GitHub Actions?

Reply:
My initial idea is to write the generator test file and test file first. The test content should be "comparison function" and "comparison parameter (data size, number of queries)", and then write a correct linear search method and faiss The same method is used to compare correctness and execution time, but since it has not yet been determined, I did not write it down.

Reply:
Do you need a large data to test the execution time? It's not a good idea to put large data into your repository since you want to test it on GitHub Action.

Reply:
This is a good question. I have thought about whether to store test data on haggingface, but if the data amount is small enough and the difference can be seen (>1s), then I think it is appropriate to use github action directly for convenience.

Reply:
OK. IMO, it still not a good choice to do benchmark in GitHub Actions since you need to deal with big data and it may not proper in GitHub Actions. Since I'm not try to place big data in GitHub Actions, I don't know is it have any disk limit or download limit strict on workflow. Please reconsider that.